bindings/go/input/dir/blob.go

```diff
@@ -0,0 +1,279 @@
+// TODO: limit recursion depth to avoid stack overflow.
+func walkDirContents(ctx context.Context, currentDir string, baseDir string,
+	opt *v1.Dir, fileSystem filesystem.FileSystem, tw *tar.Writer,
+) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("context cancelled while processing directory %q: %w", currentDir, ctx.Err())
+	default:
+	}
+
+	// Read directory contents.
+	dirEntries, err := fileSystem.ReadDir(currentDir)
+	if err != nil {
+		return fmt.Errorf("failed to read directory entries for directory %q: %w", currentDir, err)
+	}
+
+	// Iterate over directory entries.
+	for _, entry := range dirEntries {
+		// Get FileInfo for the entry.
+		fi, err := entry.Info()
+		if err != nil {
+			return fmt.Errorf("failed to get information for file %q: %w", entry.Name(), err)
+		}
+
+		// Construct the relative path of the entry with respect to the base directory.
+		entryPath := filepath.Join(currentDir, entry.Name())
+		// Check, if the entry should be included in the tar archive.
+		include, err := isPathIncluded(entryPath, opt.ExcludeFiles, opt.IncludeFiles)
+		if err != nil {
+			return fmt.Errorf("failed to check if entry %q should be included in the tar archive: %w", entryPath, err)
+		}
+		if !include {
+			continue
+		}
+
+		// Create tar header.
+		header, err := createTarHeader(fi, "", opt.Reproducible)
+		if err != nil {
+			return fmt.Errorf("failed to create tar header for file %q: %w", fi.Name(), err)
+		}
+		// Set header name to the relative path of the entry with respect to the base directory,
+		// to preserve the subfolder structure in the tar archive.
+		header.Name = entryPath
+
+		switch {
+		case entry.Type().IsRegular():
+			// The entry is a regular file.
+			// Write the header to the tar archive.
+			if err := tw.WriteHeader(header); err != nil {
+				return fmt.Errorf("failed to write tar header to tar archive: %w", err)
+			}
+
+			// Copy file content to the tar archive.
+			file, err := fileSystem.OpenFile(entryPath, os.O_RDONLY, 0o644)
+			if err != nil {
+				return fmt.Errorf("failed to open file %q: %w", entryPath, err)
+			}
+			_, err = io.CopyN(tw, file, header.Size)
+			if err != nil {
+				err = errors.Join(err, file.Close())
+				return fmt.Errorf("failed to write file %q to tar archive: %w", entryPath, err)
+			}
+			if err := file.Close(); err != nil {
+				return fmt.Errorf("unable to close file %q: %w", entryPath, err)
+			}
+
+		case entry.IsDir():
+			// The entry is a subdirectory.
+			// Write the header to the tar archive.
+			if err := tw.WriteHeader(header); err != nil {
+				return fmt.Errorf("failed to write tar header to tar archive: %w", err)
+			}
+
+			// Process subdirectory contents.
+			if err := walkDirContents(ctx, entryPath, baseDir, opt, fileSystem, tw); err != nil {
+				return err
+			}
+
+		case header.Typeflag == tar.TypeSymlink:
+			/*// The entry is a symlink.
+			if !opt.FollowSymlinks {
```

Comment:
Would be cool if you could add a TODO here with your name on the impact of this change